JaidenAshmore/java-dynamic-sqs-listener
# NONE: Added a core implementation with the basic architecture for this tool

This adds the core architecture for the SQS Listener which should hopefully provide the ability to easily configure and make it dynamic. Still got work to do but first stage.

### How to review?
1. Start at the root README.md to get a basic understanding of the architecture.
1. Read the API interfaces for futher information about the architecture
1. Download the code and test out one of the core examples, see /Users/jashmore/dev/java-dynamic-sqs-listener/examples/java-dynamic-sqs-listener-core-examples/src/main/java/com/jashmore/sqs/examples/ConcurrentBrokerExample.java which will run with a local elasticmq server.
1. Look at the core implementations.

### What is weird?
- The PrefetchingMessageRetriever has a weird way using a blocking queue to cache the prefetched messages, it works but is a little bit unintuitive.
- IndividualMessageRetriever has a weird way to handle timeouts for retrieving a message where tries to get a message every 20 seconds and will try to get the message in the provided timeout as close as possible but isn't great.
- I only just really learnt about interrupting threads so the way I implemented stopping the brokers changed a lot during the implementation, this may still be weird.
- Testing concurrent code is far more difficult.

### Next up
- Adding the Spring Starter wrapper around this. I had originally been building this at the same time but that just caused problems with the design of the API and trying to do too much at once. Now that I have the core implementation I can then create any specific classes that are needed for spring in the spring starter and not have that pollute the core implement. I had begun working this when I realised how I handle shutting down the listeners wasn't optimal so had to go and rewrite the shutting down implementation using interrupts. See https://github.com/JaidenAshmore/java-dynamic-sqs-listener/pull/3 for what I am working on in regards to this
- Add a batching message retriever that is in between the prefetching and individual message retriever. This will try and reduce the overhead of the individual message retriever that requests a message for every request but doesn't prefetch any messages and therefore won't have any messages visibility timing out before they are even processed. When a message is requested it will wait a certain period for other threads to request messages and then send a single request for multiple messages. Therefore if 7 threads are all wanting messages it will send a single call requesting for 7 messages.
- Experiment with RxJava and see if it makes the broker easier to implement with this. This will be kept outside of core so that consumers will only pull in the RxJava implementation when they need it. We don't want Core polluted.
